code/game/machinery/bloodbankgen.dm

```diff
@@ -0,0 +1,335 @@
+/obj/machinery/bloodbankgen
+	name = "blood bank generator"
+	desc = "Generates universally applicable synthetics for all blood types. Add regular blood to convert."
+	icon = 'icons/obj/bloodbank.dmi'
+	icon_state = "bloodbank-off"
+	density = TRUE
+	use_power = IDLE_POWER_USE
+	idle_power_usage = 80
+	circuit = /obj/item/circuitboard/machine/bloodbankgen
+	interaction_flags_machine = INTERACT_MACHINE_OPEN | INTERACT_MACHINE_ALLOW_SILICON
+	var/draining = FALSE
+	var/filling = FALSE
+	var/obj/item/reagent_containers/blood/bag = null
+	var/obj/item/reagent_containers/blood/outbag = null
+	var/bloodstored = 0
+	var/maxbloodstored = 1000
+	var/menustat = "menu"
+	var/efficiency = 0
+	var/productivity = 0
+
+/obj/machinery/bloodbankgen/Initialize()
+	. = ..()
+	create_reagents(1000)
+	update_icon()
+
+/obj/machinery/bloodbankgen/Destroy()
+	QDEL_NULL(bag)
+	QDEL_NULL(outbag)
+	return ..()
+
+/obj/machinery/bloodbankgen/contents_explosion(severity, target)
+	..()
+	if(bag)
+		bag.ex_act(severity, target)
+	if(outbag)
+		outbag.ex_act(severity, target)
+
+/obj/machinery/bloodbankgen/handle_atom_del(atom/A)
+	..()
+	if(A == bag)
+		bag = null
+		update_icon()
+		updateUsrDialog()
+	if(A == outbag)
+		outbag = null
+		update_icon()
+		updateUsrDialog()
+
+/obj/machinery/bloodbankgen/RefreshParts()
+	var/E = 0
+	var/P = 0
+	for(var/obj/item/stock_parts/manipulator/M in component_parts)
+		E += M.rating
+	efficiency = E
+	productivity = P
+
+/obj/machinery/bloodbankgen/update_icon()
+	cut_overlays()
+	if(is_operational())
+		icon_state = "bloodbank-on"
+
+	if(panel_open)
+		add_overlay("bloodbank-panel")
+
+	if(src.bag)
+		add_overlay("bloodbag-input")
+		if(bag.reagents.total_volume)
+			var/mutable_appearance/filling_overlay = mutable_appearance(icon, "input-reagent")
+
+			var/percent = round((bag.reagents.total_volume / bag.volume) * 100)
```

Comment:
Unresolving since this is worth consideration. Replacing this with a oneline `icon_state = "iconstate[FLOOR(round(cap/value), 10)"` would do wonders to help improve the code quality here